proxy_map.go

```diff
@@ -0,0 +1,1023 @@
+func (m *Map) TryLock(key interface{}) (bool, error) {
+	return m.tryLock(key, 0, 0)
+}
+
+// TryLockWithLease tries to acquire the lock for the specified key.
+// Lock will be released after lease time passes.
+func (m *Map) TryLockWithLease(key interface{}, lease time.Duration) (bool, error) {
+	return m.tryLock(key, lease.Milliseconds(), 0)
+}
+
+// TryLockWithTimeout tries to acquire the lock for the specified key.
+// The current thread becomes disabled for thread scheduling purposes and lies
+// dormant until one of the followings happens:
+// - The lock is acquired by the current thread, or
+// - The specified waiting time elapses.
+func (m *Map) TryLockWithTimeout(key interface{}, timeout time.Duration) (bool, error) {
+	return m.tryLock(key, 0, timeout.Milliseconds())
+}
+
+// TryLockWithLeaseTimeout tries to acquire the lock for the specified key.
+// The current thread becomes disabled for thread scheduling purposes and lies
+// dormant until one of the followings happens:
+// - The lock is acquired by the current thread, or
+// - The specified waiting time elapses.
+// Lock will be released after lease time passes.
+func (m *Map) TryLockWithLeaseTimeout(key interface{}, lease time.Duration, timeout time.Duration) (bool, error) {
+	return m.tryLock(key, lease.Milliseconds(), timeout.Milliseconds())
+}
+
+// TryPut tries to put the given key and value into this map and returns immediately.
+func (m *Map) TryPut(key interface{}, value interface{}) (bool, error) {
+	return m.tryPut(key, value, 0)
+}
+
+// TryPut tries to put the given key and value into this map and waits until operation is completed or timeout is reached.
+func (m *Map) TryPutWithTimeout(key interface{}, value interface{}, timeout time.Duration) (bool, error) {
+	return m.tryPut(key, value, timeout.Milliseconds())
+}
+
+// TryRemove tries to remove the given key from this map and returns immediately.
+func (m *Map) TryRemove(key interface{}) (interface{}, error) {
+	return m.tryRemove(key, 0)
+}
+
+// TryRemoveWithTimeout tries to remove the given key from this map and waits until operation is completed or timeout is reached.
+func (m *Map) TryRemoveWithTimeout(key interface{}, timeout time.Duration) (interface{}, error) {
+	return m.tryRemove(key, timeout.Milliseconds())
+}
+
+// Unlock releases the lock for the specified key.
+func (m *Map) Unlock(key interface{}) error {
+	if keyData, err := m.validateAndSerialize(key); err != nil {
+		return err
+	} else {
+		refID := m.refIDGenerator.NextID()
+		request := codec.EncodeMapUnlockRequest(m.name, keyData, m.lockID, refID)
+		_, err = m.invokeOnKey(m.ctx, request, keyData)
+		return err
+	}
+}
+
+// ListenEntryNotification adds a continuous entry listener to this map.
+func (m *Map) ListenEntryNotification(config MapEntryListenerConfig, handler EntryNotifiedHandler) (string, error) {
+	flags := makeListenerFlags(&config)
+	subscriptionID := int(m.refIDGenerator.NextID())
```

Comment:
I hesitate to use the UUID type in the public API, since that would force the user to depend on our internal UUID implementation without much benefit. I've changed event subscription ID type to int64

Reply:
A string may be returned back to the user instead of the UUID type. Also, all Hz v4.0 clients need to support UUID serialization, so a dependency or a in-house implementation will have to be added into the Go client anyway.

Reply:
Another problem with the current approach is collision of subscription ids between different `Map`s. Since each `Map` gets a new `ReferenceIDGenerator`, assigned subscription ids in all maps start with 1 and overwrite each other when `m.userEventDispatcher.Subscribe()` call happens. This could be mitigated by using UUIDs (since they provide good level of entropy) or by using a global, per-client `ReferenceIDGenerator` for all user subscriptions.

Reply:
Refactored listeners to use UUIDs: https://github.com/yuce/hazelcast-go-client/commit/15136dbdd26d8f9e1acdf9eba6c5ef62be6392d4